controllers/machinehealthcheck_targets_test.go

```diff
@@ -0,0 +1,380 @@
+			toCreate:        baseObjects,
+			expectedTargets: nil,
+		},
+		{
+			desc:     "when a machine's node is missing",
+			toCreate: append(baseObjects, testMachine1),
+			expectedTargets: []healthCheckTarget{
+				{
+					Machine: testMachine1,
+					MHC:     testMHC,
+					Node: &corev1.Node{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: "node1",
+						},
+					},
+				},
+			},
+		},
+		{
+			desc:     "when a machine's labels do not match",
+			toCreate: append(baseObjects, testMachine1, testMachine2, testNode1),
+			expectedTargets: []healthCheckTarget{
+				{
+					Machine: testMachine1,
+					MHC:     testMHC,
+					Node:    testNode1,
+				},
+			},
+		},
+		{
+			desc:     "with multiple machines, should match correct nodes",
+			toCreate: append(baseObjects, testNode1, testMachine1, testNode3, testMachine3, testNode4, testMachine4),
+			expectedTargets: []healthCheckTarget{
+				{
+					Machine: testMachine1,
+					MHC:     testMHC,
+					Node:    testNode1,
+				},
+				{
+					Machine: testMachine3,
+					MHC:     testMHC,
+					Node:    testNode3,
+				},
+				{
+					Machine: testMachine4,
+					MHC:     testMHC,
+					Node:    testNode4,
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			gs := NewGomegaWithT(t)
+
+			gs.Expect(clusterv1.AddToScheme(scheme.Scheme)).To(Succeed())
+			k8sClient := fake.NewFakeClientWithScheme(scheme.Scheme, tc.toCreate...)
+
+			// Create a test reconciler
+			reconciler := &MachineHealthCheckReconciler{
+				Client: k8sClient,
+				Log:    log.Log,
+				scheme: scheme.Scheme,
+			}
+
+			targets, err := reconciler.getTargetsFromMHC(k8sClient, testCluster, testMHC)
+			gs.Expect(err).ToNot(HaveOccurred())
+
+			gs.Expect(targets).To(ConsistOf(tc.expectedTargets))
+		})
+	}
+}
+
+func TestHealthCheckTargets(t *testing.T) {
+	namespace := "test-mhc"
+	clusterName := "test-cluster"
+	mhcSelector := map[string]string{"cluster": clusterName, "machine-group": "foo"}
+
+	// Create a test MHC
+	testMHC := &clusterv1.MachineHealthCheck{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-mhc",
+			Namespace: namespace,
+		},
+		Spec: clusterv1.MachineHealthCheckSpec{
+			Selector: metav1.LabelSelector{
+				MatchLabels: mhcSelector,
+			},
+			ClusterName: clusterName,
+			UnhealthyConditions: []clusterv1.UnhealthyCondition{
+				{
+					Type:    corev1.NodeReady,
+					Status:  corev1.ConditionUnknown,
+					Timeout: metav1.Duration{Duration: 5 * time.Minute},
+				},
+				{
+					Type:    corev1.NodeReady,
+					Status:  corev1.ConditionFalse,
+					Timeout: metav1.Duration{Duration: 5 * time.Minute},
+				},
+			},
+		},
+	}
+
+	testMachine := newTestMachine("machine1", namespace, clusterName, "node1", mhcSelector)
+
+	// Target for when the node has not yet been seen by the Machine controller
+	testMachineLastUpdated400s := testMachine.DeepCopy()
+	nowMinus400s := metav1.NewTime(time.Now().Add(-400 * time.Second))
```

Comment:
Is there any chance for test flakes using a real clock here and in the controller? apimachinery has a `Clock` interface with `RealClock` and `FakeClock` implementations.

Reply:
I deliberately set these as quite large times expecting that each test case should really take no more than a couple of seconds. There _could_ potentially be some flakiness from this but that would involve individual test cases hanging for minutes right? I'd be tempted to leave it for now and follow up with an issue if it becomes one? Is that ok?

Reply:
I wrote this before I saw you were truncating seconds. I think if we use the fake clock we wouldn't need that, as we'd be in strict control. But yes 👍 we can leave as-is and fix later if we see flakes.